sw/device/lib/dif/dif_spi_device.c

```diff
@@ -0,0 +1,695 @@
+ *
+ * If `is_recv` is set, then the copy direction is `spi -> buf`. If it is
+ * unset, the copy direction is `buf -> spi`.
+ *
+ * @param spi a SPI device.
+ * @param fifo a decompressed FIFO pointer pair.
+ * @param fifo_base the base address in SRAM for the FIFO to copy to/from.
+ * @param fifo_len the length of the FIFO, in bytes.
+ * @param byte_buf a main memory buffer for copying from/to.
+ * @param buf_len the length of the main memory buffer.
+ * @param is_recv whether this is a SPI reciept or SPI transmit transaction.
+ * @return the number of bytes copied.
+ */
+static size_t spi_memcpy(const dif_spi_device_t *spi, fifo_ptrs_t *fifo,
+                         uint16_t fifo_base, uint16_t fifo_len,
+                         uint8_t *byte_buf, size_t buf_len, bool is_recv) {
+  uint16_t bytes_left = fifo_bytes_in_use(*fifo, fifo_len);
+  // When sending, the bytes left are the empty space still available.
+  if (!is_recv) {
+    bytes_left = fifo_len - bytes_left;
+  }
+
+  if (bytes_left > buf_len) {
+    bytes_left = buf_len;
+  }
+  if (bytes_left == 0) {
+    return 0;
+  }
+
+  // For receipt, we advance the read pointer, which indicates how far ahead
+  // we've read so far. For sending, we advance the write pointer, which
+  // indicates how far ahead we've written.
+  fifo_ptr_t *ptr;
+  if (is_recv) {
+    ptr = &fifo->read_ptr;
+  } else {
+    ptr = &fifo->write_ptr;
+  }
+
+  uint16_t total_bytes_available = bytes_left;
+
+  // First, bring the fifo write pointer into word alignment, so we can do
+  // simple full-word I/O rather than partial word I/O; note that the FIFO
+  // SRAM is not guaranteed to support subword writes.
+  ptrdiff_t misalignment = misalignment32_of(ptr->offset);
+  if (misalignment != 0) {
+    // The number of bytes missing to bring `ptr->offset` back into alignment.
+    // For exmaple, 0x3 has misalignment of 3 and realignment of 1.
+    ptrdiff_t realignment = sizeof(uint32_t) - misalignment;
+    // Note that we might be doing less I/O than the misalignment requires; we
+    // might be off by a single byte, but not have the full three bytes for full
+    // realignment.
+    if (realignment > bytes_left) {
+      realignment = bytes_left;
+    }
+
+    // Converts `offset`, which points to a subword boundary, to point to the
+    // start of the current word it points into.
+    ptrdiff_t current_word_offset = ptr->offset - misalignment;
+    uint32_t current_word =
+        mmio_region_read32(spi->base_addr, fifo_base + current_word_offset);
+
+    // Act on only to a suffix of `current_word`, corresponding to the necessary
+    // realignment.
+    uint8_t *current_subword = ((uint8_t *)&current_word) + misalignment;
+    if (is_recv) {
+      memcpy(byte_buf, current_subword, realignment);
+    } else {
+      // When sending, we need to commit the staged write.
+      memcpy(current_subword, byte_buf, realignment);
+      mmio_region_write32(spi->base_addr, fifo_base + current_word_offset,
```

Comment:
Reading the comments of this function and looking at the variable name -`dif_spi_device_t` has fields with similar names- I thought `fifo_base` was relative to the SRAM. But it is relative to the device base addr. Moving `SPI_DEVICE_BUFFER_REF_OFFSET +` here can help make things clearer.

Reply:
Done.